app/javascript/lib/scatter-plot.js

```diff
@@ -0,0 +1,308 @@
+    Plotly.update(target, dataUpdate)
+  })
+}
+
+/** Get Plotly layout object for scatter plot */
+export function getBaseLayout(height, width) {
+  const font = labelFont
+
+  const layout = {
+    hovermode: 'closest',
+    margin: {
+      t: 25,
+      r: 0,
+      b: 20,
+      l: 0
+    },
+    height,
+    width,
+    font
+  }
+  return layout
+}
+
+/** Gets Plotly layout scene props for 3D scatter plot */
+export function get3DScatterProps(camera, cluster) {
+  const { domainRanges, axes } = cluster
+  const { titles, ranges, aspects } = axes
+
+  const scene = {
+    camera,
+    aspectmode: 'cube',
+    xaxis: { title: titles.x, autorange: true, showticklabels: false },
+    yaxis: { title: titles.y, autorange: true, showticklabels: false },
+    zaxis: { title: titles.z, autorange: true, showticklabels: false }
+  }
+
+  if (domainRanges) {
+    scene.xaxis.autorange = false
+    scene.xaxis.range = ranges.x
+    scene.yaxis.autorange = false
+    scene.yaxis.range = ranges.y
+    scene.zaxis.autorange = false
+    scene.zaxis.range = ranges.x
+    scene.aspectmode = aspects.mode,
+    scene.aspectratio = {
+      x: aspects.x,
+      y: aspects.y,
+      z: aspects.z
+    }
+  }
+
+  return scene
+}
+
+/** Gets Plotly layout props for 2D scatter plot */
+export function get2DScatterProps(cluster) {
+  const {
+    axes, domainRanges, hasCoordinateLabels, coordinateLabels
+  } = cluster
+  const { titles } = axes
+
+  const layout = {
+    xaxis: { title: titles.x, showticklabels: false },
+    yaxis: { title: titles.y, showticklabels: false, scaleanchor: 'x' }
+  }
+
+  // if user has supplied a range, set that, otherwise let Plotly autorange
+  if (domainRanges) {
+    layout.xaxis.range = domainRanges.x
+    layout.yaxis.range = domainRanges.y
+  } else {
+    layout.xaxis.autorange = true
+    layout.yaxis.autorange = true
+  }
+
+  if (hasCoordinateLabels) {
+    layout.annotations = coordinateLabels
+  }
+
+  return layout
+}
+
+/** Set colors for group annotations */
+export function setMarkerColors(data) {
+  return data.map((trace, i) => {
+    trace.marker.color = getColorBrewerColor(i)
+    return trace
+  })
+}
+
+/** Get height and width for a to-be-rendered cluster plot */
+function calculatePlotRect(numRows, numColumns) {
+  // Get height
+  const $ideo = $('#_ideogram')
+  const ideogramHeight = $ideo.length ? $ideo.height() : 0
+  const verticalPad = 225 // Accounts for page header, search area, nav tabs
+  let height = $(window).height() - verticalPad - ideogramHeight
+  if (numRows > 1) {height = height/numRows - 10}
+
+  // Get width, and account for expanding "View Options" after page load
+  const baseWidth = $('#render-target .tab-content').actual('width')
+  const horizontalPad = 80 // Accounts for empty space to left and right
+  let width = (baseWidth - horizontalPad) / numColumns
+
+  // Ensure plots aren't too small
+  if (height < 100) {height = 100}
+  if (width < 100) {width = 100}
```

Comment:
I think using Math.max is the more conventional way to set min values like this, but either way is fine
```suggestion
  height = Math.max(height, 100)
  width = Math.max(width, 100)
```